### PR TITLE
feat: Add initial support for building RPMs using Packit

### PR DESCRIPTION
**Pull request by jirihnidek**
* Card ID: CCT-231
* Added .packit.yml
* RPMs are build, when PR is created/updated
* RPMs are build for CentOS Strem 8/9, RHEL 8/9 and all Fedoras
* Added autochangelog and set source_date_epoch_from_changelog to 0.
  Use %autochangelog only for rhel9+ and fedora.
* Deleted .copr/Makefile
* Moved rhc.spec.in to ./
* Added rhc.spec target to Makefile and modified this file little bit

### CODE REVIEW

**Comment by ahitacat** [comment]
You removed the dependency on go. Maybe try to run the dnf as a part of the srpm build. Like in [yggdrasil](https://github.com/RedHatInsights/yggdrasil/blob/main/.copr/Makefile#L17) project.

**Comment by jirihnidek** [comment]
It is possible to buld RPMs using packit and COPR too:

https://copr.fedorainfracloud.org/coprs/g/yggdrasil/latest/build/6950757/

File: .copr/Makefile
Comment (ptoscano) [inline .copr/Makefile:54]: while I can see what you want to achieve, IMHO this makes things more complicated:
- in addition to the package list in the `dnf install` command and in the `srpm_build_deps` list of the packit config, there are more package names in the `grep` commands
- running `rpm -qa` is slow...

Since all we need there are the spec file and the source tarball, I think the build can be simplified a lot:
- in `.copr/Makefile`:
   - move the spec file generation in its own target (eg `spec`), adding it as dependency of the `srpm` target; this would be the `sed` call
   - move the tarball generation in its own target (eg `tarball`), adding it as dependency of the `srpm` target; this would be only the `make dist` call, no need to also copy the tarball anywhere
- in the packit config:
   - run `make -f .copr/Makefile spec` as only step in `post-upstream-clone`
   - run `make -f .copr/Makefile tarball` as first step in `create-archive`
- drop `rpm-build` from `srpm_build_deps`, as it will not be needed anymore

**Review by ptoscano — CHANGES_REQUESTED** [review]

File: .packit.yml
Comment (jirihnidek) [reply .packit.yml]: It seems that you are right, but it fails when I try to run the command without a shell. It seems that it tries to use wrong directory for running the command.

File: .copr/Makefile
Comment (jirihnidek) [reply .copr/Makefile]: It just tried to use `"yggdrasil"` and not `yggdrasil` string for some actions and caused various issues.

File: .copr/rhc.spec.in
Comment (jirihnidek) [reply .copr/rhc.spec.in:79]: I didn't know that I can cherry-pick anything from other repository. I will give a try.

File: .copr/Makefile
Comment (jirihnidek) [reply .copr/Makefile]: It was little bit more complicated, but I use proposed approach.

File: .copr/rhc.spec.in
Comment (jirihnidek) [reply .copr/rhc.spec.in:79]: I can see that it is possible now, but I don't want to do such thing only because it is possible.

**Comment by jirihnidek** [comment]
COPR build works again using this branch: https://copr.fedorainfracloud.org/coprs/g/yggdrasil/latest/build/7048982/

File: .copr/rhc.spec.in
Comment (ptoscano) [reply .copr/rhc.spec.in:79]: I'm not sure I understand what you said.

The changelog is generated automatically, hence a static version a) is not useful b) would get stale pretty soon. Keeping it the sources would only be a waste, and I do not see how/what it would do good.

File: .copr/Makefile
Comment (ptoscano) [inline .copr/Makefile:26]: While I understand the idea behind this, I'd rather keep `make srpm` for copr only, and thus here unconditionally installing the needed packages with `dnf`. Overloading `make srpm` with other meanings than what copr wants means complicating this `Makefile`, and the code above shows exactly this.

With the new split targets for the spec file and the tarball, now there is less need to use `make srpm` manually. If it is wanted to have also an easy way to generate the srpm, I'd simply add a new target:
```make
source-rpm: spec tarball
    rpmbuild -bs $(specdir)/rhc.spec

srpm: deps source-srpm
    ...
```
this would have the advantages of reusing the existing step (part of the `make srpm` flow for copr), and not overload it with other meanings. Bonus point: using this new target would not bloat your local "SOURCES" and "SRPMS" directories of rpmbuild.

File: .packit.yml
Comment (ptoscano) [inline .packit.yml:18]: minor nit: you can call two makefile targets at the same time:
```suggestion
    - bash -c 'make -f .copr/Makefile spec tarball'
```

**Review by ptoscano — CHANGES_REQUESTED** [review]
Generally good; two more things to improve:
- drop the static RPM changelog
- avoid overloading the `make srpm` target for copr

File: .copr/Makefile
Comment (jirihnidek) [reply .copr/Makefile:26]: When there is empty `%changelog` or no `%changelog` in `rhc.spec` file, then warning is printed during linting of rpm.

I don't like idea of having two targets for building `src.rpm`. Current solution just works. `make srpm` makes `src.rpm`. There is no overloading of meaning.

Comment (ptoscano) [reply .copr/Makefile:26]: > When there is empty `%changelog` or no `%changelog` in `rhc.spec` file, then warning is printed during linting of rpm.

That warning is harmless. If that is a concern, I'd rather use `%autochangelog` instead; please do not store a static changelog that is useless and gets obsolete soon.

> I don't like idea of having two targets for building `src.rpm`. Current solution just works. `make srpm` makes `src.rpm`.

`make srpm` also copies the bits to the "SOURCES" and "SRPMS" directories of rpmbuild (see the output of `rpmbuild -E %_sourcedir` and `rpmbuild -E %_srcrpmdir`), because that is the "interface" of `make srpm` for copr. Copying them for a local run is wasteful, because it fills up your own local directories with tarballs and srpm files.

> There is no overloading of meaning.

There is: now `make srpm` has to work also for what was not designed (i.e. build in copr).

Comment (ptoscano) [reply .copr/Makefile:26]: > I really don't care about copying files to "SOURCES" and "SRPMS".

That isn't really a reason to silently fill up some place that you might not even know about.

> We all can use `rm` command.

Let's go by your logic then: we all can use `rpmbuild` command.

> Does `make srpm` work? It does. There is no problem.

`make srpm` is for copr, please leave it to copr exclusively without overloading its function. The fact that it "works for you" does not mean it is a good idea in general or there is no problem with it. Implementing an hack and saying that "it works" is not a good way forward.

Please drop this hack, thanks, as already asked.

Comment (jirihnidek) [reply .copr/Makefile:26]: The `make srpm` is common target in other projects e.g. https://github.com/candlepin/virt-who/blob/main/Makefile#L16. The necessity to run `make srpm` as a root is a hack. The necessity of installing any RPM as a side effect of `srpm` target is a hack. I try to make it only less painful.

BTW: I would say that you are not contributor to this project and your are not in the position to request or enforce anything. Your advice is welcome, but the style of communication you are showing in this PR is not acceptable.

Comment (subpop) [reply .copr/Makefile:26]: The `srpm` target in the file `.copr/Makefile` is an interface for building an SRPM required by COPR. Hypothetically, if COPR were to change the required output of the `make srpm` interface, we would be compelled to update the `srpm` target follow the new expectations, or our system would stop working with COPR. To me, that means that `make srpm` is not a general-purpose "build a source RPM" make recipe. Instead, it is a COPR-specific interface. Additionally, I think COPR assumes that it always invokes `make -f .copr/Makefile srpm` as root. One more reason why this interface is very COPR-specific.

We would be wise to abstract ourselves away from having to comply with changes that might arise due to that external API. (I did this in yggdrasil; a general-purpose SRPM build recipe exists in the `dist/srpm` directory, and COPR's `.copr/Makefile` interface uses it.) If we want to support a general-purpose SRPM build recipe, it should be in `./Makefile` instead of `.copr/Makefile`. virt-who appears to do it this way; it has a general, top-level `Makefile` with a 'srpm' target that invokes `tito` to compile a source RPM.

Comment (jirihnidek) [reply .copr/Makefile:26]: COPR can assumes that `make -f .copr/Makefile srpm` is invoked by root. Current change will not influence this assumption.

If COPR build was broken, then fixing/testing would be easier with this change... It would allow to trigger `make -f .copr/Makefile srpm` on local machine as non-root user. I think that it is better to try to fix/test broken COPR build locally as non-root user, then running something like this as root, or re-trigger COPR build again and again after change is pushed to some PR.

I aware that this change is not perfect... I would like to improve it (e.g. it would be good to have the list of required packages in some variable and do not duplicate code), but first I need to know that something like this is wanted.

Comment (subpop) [reply .copr/Makefile:26]: Good points. I agree that it does make debugging the COPR makefile and target easier than it is now.

**Review by subpop — APPROVED** [review]

**Comment by subpop** [comment]
I implemented a Packit-only solution in yggdrasil that works for the `yggdrasil-0.2` branch: https://github.com/RedHatInsights/yggdrasil/pull/219. This approach eliminates the redundancy we're carrying between COPR and Packit, in favor of Packit exclusively. We should be able to take a similar approach for `rhc` too.

File: Makefile
Comment (subpop) [inline Makefile:32]: This VERSION is overriding the VERSION previously defined in the Makefile.

File: Makefile
Comment (subpop) [inline Makefile:33]: I don't customize the VERSION and RELEASE variables in the yggdrasil-0.2 implementation. When packit sends a build to COPR, it modifies the Release field of the spec file, encoding the commit information.

**Review by subpop — CHANGES_REQUESTED** [review]
The implementation I made for yggdrasil-0.2 uses a pattern rule to create a file from a .in file. This implicitly works for the template file `yggdrasil.spec.in`, without needing to explicitly include anything RPM-related in the Makefile.

File: Makefile
Comment (subpop) [inline Makefile:165]: You can then replace these named paths with `make` references:

```suggestion
		$^ > $@
```

I suggest you consider making this a pattern rule. It can turn a single "rhc.spec from rhc.spec.in" rule into a general-purpose "any_file from any_file.in" rule.

File: .packit.yml
Comment (subpop) [inline .packit.yml:39]: Does this additional repo need to be here when you're already building into that same repo?

**Review by subpop — CHANGES_REQUESTED** [review]

File: .packit.yml
Comment (jirihnidek) [reply .packit.yml:39]: For some reason we have `BuildRequires: yggdrasil >= 0.4` in `rhc.spec.in` file. Is it really necessary for upstream repository?

File: Makefile
Comment (jirihnidek) [reply Makefile:146]: We use `spec` rule in other projects (e.g. in yggdrasil). I would like to be consistent.

File: .packit.yml
Comment (jirihnidek) [reply .packit.yml:39]: Can we try to fix this in another PR? It seems that yggdrasil is not needed for building `rhc`? I don't want to mix too many things together.

Comment (subpop) [reply .packit.yml:39]: Yea, that's fair. Let's clean this up in a followup PR.

**Comment by jirihnidek** [comment]
@subpop is there anything blocking merging this PR? I think Pino's request is irrelevant since support for COPR was dropped.

**Pull request merged**
